File: 093. Restore IP Addresses/C++/Solution.cpp
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        class Utils {
        public:
            void backTrack(string &s, vector<string> &result, string &current, int index, int count) {
                if (count > 4) {
                    return;
                }
                if (count == 4 && index == s.length()) {
                    result.push_back(current.substr(0, current.length() - 1));
                } else if (count < 4) {
                    for (int i = 0; i < 3; i++) {
                        if (index + i >= s.length()) {
                            break;
                        }
                        string sub = s.substr(index, i + 1);
                        if (stoi(sub) >= 256 || (sub[0] == '0' && sub.length() > 1)) {
                            continue;
                        }
                        string temp = current + sub + ".";
                        backTrack(s, result, temp, index + i + 1, count + 1);
                    }
                }
            }
        };
        vector<string> result;
        Utils utils;
        string temp = "";
        utils.backTrack(s, result, temp, 0, 0);
        return result;
    }
};

File: 093. Restore IP Addresses/Python/Solution.py
class Solution(object):
    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        def back_trace(s, result, current, index, count):
            if count > 4:
                return
            if count == 4 and index == len(s):
                result.append(current[:-1])
            elif count < 4:
                for i in range(3):
                    if index + i >= len(s):
                        break
                    sub = s[index:index + i + 1]
                    if int(sub) >= 256 or (sub[0] == '0' and len(sub) > 1):
                        continue
                    back_trace(s, result, current + sub + '.', index + i + 1, count + 1)

        result = []
        back_trace(s, result, '', 0, 0)
        return result
